segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadOperation.java

```diff
@@ -132,7 +133,9 @@ private void logEnd() {
                             return txn.get(currentChunkName)
                                     .thenAcceptAsync(storageMetadata -> {
                                         chunkToReadFrom = (ChunkMetadata) storageMetadata;
-                                        log.debug("{} read - reading from next chunk - segment={}, chunk={}", chunkedSegmentStorage.getLogPrefix(), handle.getSegmentName(), chunkToReadFrom);
+                                        Preconditions.checkState(null != chunkToReadFrom, "chunkToReadFrom is null");
```

Comment:
Maybe not related to the PR, but the way `chunkToReadFrom` (and other variables) is set and checked here and there looks hard to track in case we face a race conditions. Even tough a variable is declared as `volatile`, the point is that if multiple methods (perhaps executed by different threads) are setting and reading this variable throughout the class, it increases considerably the complexity of debugging it in case of problems. Perhaps this was the only way, I don't know, but maybe in the future we can reconsider this pattern to see if there is an alternative that simplifies the code.

Reply:
Generic answer to your question - 
Unfortunately java does not have [Async/Await syntax suger](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/async/) that generates all the future based  ([Task parallel library](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/task-parallel-library-tpl)) lambdas for your code.

So we have to do it by hand. Having done lots TPL based programming in the past, I was sort of reluctant to rewrite this code for async knowing it makes very difficult to maintain spaghetti code with lots of lambdas. 
However we absolutely do need it here.

The general pattern that I have here for thread safety is as follows
- Each major function in linear non-async code becomes a _method object_. Each invocation is a separate object. So separate invocations share nothing.
- All members **must** be either **final** or **volatile**
- If you need to use any arithmetic then it is better to use Atomic variables (which actually just volatile variables with additional safety added for atomic addition/subtraction
- The lamdas are executed in series by java sdk (i.e CompletableFuture framework) one lambda at a time.
- The JVM [memory model](https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html) guarantees that reads and writes to volatile variable are thread safe.

Writing all this async code by hand and yet keeping it understandable is not easy. But hopefully code gets better, faster and also cleaner as go.